pandas/tests/frame/test_sort_values_level_as_str.py

```diff
@@ -0,0 +1,122 @@
+    ['outer', 'inner']
+])
+def df_idx(request, df_none):
+    levels = request.param
+    return df_none.set_index(levels)
+
+
+@pytest.fixture(params=[
+    'inner',     # index level
+    ['outer'],   # list of index level
+    'A',         # column
+    [('B', 5)],  # list of column
+    ['inner', 'outer'],   # two index levels
+    [('B', 5), 'outer'],  # index level and column
+    ['A', ('B', 5)],      # Two columns
+    ['inner', 'outer']    # two index levels and column
+])
+def sort_names(request):
+    return request.param
+
+
+@pytest.fixture(params=[True, False])
+def ascending(request):
+    return request.param
+
+
+def test_sort_index_level_and_column_label(
+        df_none, df_idx, sort_names, ascending):
+
+    # Get index levels from df_idx
+    levels = df_idx.index.names
+
+    # Compute expected by sorting on columns and the setting index
+    expected = df_none.sort_values(by=sort_names,
+                                   ascending=ascending,
+                                   axis=0).set_index(levels)
+
+    # Compute result sorting on mix on columns and index levels
+    result = df_idx.sort_values(by=sort_names,
+                                ascending=ascending,
+                                axis=0)
+
+    assert_frame_equal(result, expected)
+
+
+def test_sort_column_level_and_index_label(
+        df_none, df_idx, sort_names, ascending):
+
+    # Get levels from df_idx
+    levels = df_idx.index.names
+
+    # Compute expected by sorting on axis=0, setting index levels, and then
+    # transposing. For some cases this will result in a frame with
+    # multiple column levels
+    expected = df_none.sort_values(by=sort_names,
+                                   ascending=ascending,
+                                   axis=0).set_index(levels).T
+
+    # Compute result by transposing and sorting on axis=1.
+    result = df_idx.T.sort_values(by=sort_names,
+                                  ascending=ascending,
+                                  axis=1)
+
+    if len(levels) > 1:
+        # Accessing multi-level columns that are not lexsorted raises a
+        # performance warning
+        with tm.assert_produces_warning(PerformanceWarning,
+                                        check_stacklevel=False):
+            assert_frame_equal(result, expected)
+    else:
+        assert_frame_equal(result, expected)
+
+
+def test_sort_values_column_index_level_precedence():
+    # GH 14353, when a string passed as the `by` parameter
+    # matches a column and an index level the column takes
+    # precedence
+
+    # Construct DataFrame with index and column named 'idx'
+    idx = Index(np.arange(1, 7), name='idx')
+    df = DataFrame({'A': np.arange(11, 17),
+                    'idx': np.arange(6, 0, -1)},
+                   index=idx)
+
+    # Sorting by 'idx' should sort by the idx column and raise a
+    # FutureWarning
+    with tm.assert_produces_warning(FutureWarning, check_stacklevel=False):
```

Comment:
Can you see if you can get this working without `check_stacklevel=False` ?

Reply:
Sure.  Could you offer any insight into what `check_stacklevel` is actually checking? I've had trouble understanding that @jorisvandenbossche

Reply:
See https://github.com/pandas-dev/pandas/issues/9584 and https://github.com/pandas-dev/pandas/pull/10676. 
The `stacklevel` is specified in the FutureWarning in `_check_label_or_level_ambiguity`, and is currenlty set at `2`, which is not correct. This should be higher, depending on how many call steps there are between `sort_values` and calling that function. 
The main problem here will probably be that  `_check_label_or_level_ambiguity` is used in multiple places, so the needed stacklevel might differ for the different usages.

Reply:
Thanks, that's helpful

Reply:
The goal is to report the users's line of code that triggered the warning. In our code, when you emit the `FutureWarning` there's an optional `stacklevel` parameter that's controls which line of code the error message is reported with. It's supposed to be the number of function calls from user code to our method emitting the warning.

It's not always possible to get that correct, if you have multiple pandas methods calling the method emitting the warning.

Reply:
Ok, I was able to remove `check_stacklevel=False` for these tests, and for the related tests for `groupby` and `merge`.  I added a `stacklevel` parameter to `_check_label_or_level_ambiguity` and `_get_label_or_level_values` that behaves like the `stacklevel` parameter of `warnings.warn` itself (`stacklevel=1` blames the caller, `stacklevel=2` blames the callers parent, etc.).

Does this look like a good way to handle it? @TomAugspurger @jorisvandenbossche